src/throttler/throttler.ts

```diff
@@ -254,4 +239,71 @@ export abstract class Throttler<T, R> {
       return p.resolve();
     });
   }
+
+  private initializeTimeout(cursorIndex: number): Promise<void> {
+    const taskData = this.taskDataMap.get(cursorIndex)!;
+    const timeoutMillis = taskData.timeoutMillis!;
+    const timeoutPromise = new Promise<void>((_, reject) => {
+      taskData.timeoutId = setTimeout(() => {
+        taskData.isTimedOut = true;
+        reject(new TimeoutError(this.taskName(cursorIndex), timeoutMillis));
+      }, timeoutMillis);
+    });
+
+    return timeoutPromise;
+  }
+
+  private async executeTask(cursorIndex: number): Promise<any> {
+    const taskData = this.taskDataMap.get(cursorIndex)!;
+    const t0 = Date.now();
+    try {
+      const result = await this.handler(taskData.task);
+      const dt = Date.now() - t0;
+      this.min = Math.min(dt, this.min);
+      this.max = Math.max(dt, this.max);
+      this.avg = (this.avg * this.complete + dt) / (this.complete + 1);
+
+      return result;
+    } catch (err) {
+      if (taskData.retryCount === this.retries) {
+        throw new RetriesExhaustedError(this.taskName(cursorIndex), this.retries, err);
+      }
+      await backoff(taskData.retryCount + 1, this.backoff);
```

Comment:
Maybe after the timeout check.

Reply:
since running `this.executeTask` is potentially much more expensive than `backoff`, putting the check after will make sure that if it's timed out during the backoff, `this.executeTask()` will not be called again